### ISSUE TITLE
PAM/AD auth doesn't create homedir on first login

### ISSUE THREAD

**Issue by mogthesprog** [jupyter/jupyterhub#295]
So from what i can see in auth.py, the 'login' pam service is used to authenticate the user. For some reason I don't think the 'session' part of this service is honoured. I've tested this by using pamtester with the login service (see below) and the home directory is created.

pamtester login <username> open_session close_session

Any idea why this may be? I think it has something to do with the simplepam python module functionality but i'm not certain.

any help is appreciated. Cheers

**Comment by minrk** [comment]
I'm not sure what would be different, but I will try to investigate.

**Comment by mogthesprog** [comment]
Hey, chers for quick reply. Done a bit of digging and it seems like the simple pam module doesn't implement any PAM arguemnts (open/close session as stated above). I have however found a modified module here - https://github.com/gnosek/python-pam. If you take a look at pam.py, you'll notice the implementation of a handle for passing raw PAM arguments, such as open session/close session.

If these were implemented, then home directory creation of the 'login' service would be honoured.

However, I don't think this is an issue with jupyterhub as such. It seems to me like the simplepam module in python doesn't allow this functionality...

Just to give you a bit of background on my use of jupyter. I've got a container based jupyter instance that i've implemented in CentOS. Jupyter relies on the home directory for storage of some critical files so if the home directory isn't there then the server wont start after login. I've considered mounting the home directory of the host into the container, however this doesn't really solve the problem. If the use has never logged in to this machine before then they have no home directory and we have the same problem.

Any help/advice is much appreciated. I think since containers are becoming more popular now, maybe there is a need for the session information to be used? maybe I'm wrong though.
